design/MultipleQueues.md

```diff
@@ -0,0 +1,83 @@
+Queue families with `graphics` or `compute` will always have `copy`.
+There may be queue families in the future like D3D12's video decode queues, which do not support copy operations.
+If there are any families that support graphics, there is at least one family that supports both compute and graphics.
+The most capable family is always first in the `Adapter.queueFamilies` list.
+(It will either be a graphics+compute, or at least compute)
+Exposed adapters must have at least a compute family.
+Users can also create `QueueFamily`s.
+
+## Queue Creation
+
+Queues are created at Device creation time, and are exposed via `sequence<Queue> Device.queues`.
+Each element in `sequence<QueueFamily> WebGPUDeviceDescriptor.queueRequests` results in a corresponding element in `Device.queues`.
+If an user-provided QueueFamily does not match a family in `Adapter.queueFamilies`, a more capable family is used.
+If no available family can satisfy an user-provided family, device creation fails.
+
+## Synchronization
+
+Queues can have fences inserted into them, and any queue can wait on that fence to be complete.
+
+Synchronizing access to resources across queues should be done by telling a queue to wait until a fence is complete, followed by submitting command buffers that depend on that fence.
+Resource data may have multiple concurrent readers, or exclusively one writer, at a time.
+Resources may have multiple concurrent readers and writers, so long as all subranges satisfy many-read/single-write exclusion. [1]
+If different commands would violate this exclusion, the implmentation injects synchronization.
+If a command is submitted that will never be able to synchronize for exclusion without subsequent user commands, that Submit is refused.
+
+[1]: For MVP, we require reads/write exclusion at whole-resource granularity, instead of allowing subranges.
+
+### How Is Implicit Synchronization Done
+
+Implicit Synchronization is clearly possible via the degenerate approach of strict serialization.
+Viability depends on how difficult it is to inject synchronization with minimal overhead, and in particular minimal excess serialization.
+
+Each CommandBuffer knows all reads and writes to its resources, and satisfies reads/write exclusion internally.
+(out of scope of multi-queue discussion)
+Further, baked CommandBuffers know their required starting memory barrier requirements, as well as their end state.
+
+Each Resource effectively has:
+
+~~~
+struct LastAccess {
+   CommandBuffer cb;
+   AccessBits access;
+};
+list<LastAccess> last_accesses;
+~~~
+
+Each CommandBuffer effectively has:
+
+~~~
+struct ResourceAccess {
+   Resource res;
+   AccessBits src_access;
+   AccessBits dst_access;
+};
+list<ResourceAccess> res_accesses;
+~~~
+
+For a `submit(sequence<WebGPUCommandBuffer> buffers)`, implementations inject any required synchronization.
+On submit, each CommandBuffer in turn traverses its resources and identifies any outstanding synchronization requirements.
```

Comment:
👍

Reply:
wait.. this means that you will require and expect submitted command buffers to execute in-order of submission!?

Reply:
What this is saying is that we (WebGPU implementation) will insert command buffers doing the memory barriers / resource transitions between some of those in a sequence. That doesn't serialize the execution (by GPU) of the command buffers more than the user would do in Vulkan.

E.g. if one command buffer is writing into a UAV and another one is using the same resource as a vertex buffer, WebGPU implementation will insert the transition. The user would do the same in Vulkan/D3D12.

Reply:
Ok, so you will go on a quest to find dependency chains between the uses of resources?

I hate to tell you that, but it is exactly what OpenGL has to do... and that makes creating a bugless and performant OpenGL implementation (driver) very very hard.
You'd have to do amazing work on par with the engineers at Nvidia and AMD to both:
1. insert the transitions correctly to not get undefined behaviour, artefacts, other problems
2. not insert too many to get subpar performance and serrialization
3. do it fast

I think a dependency graph of the resources used in the command buffers you will be processing for submission will be an enormous DAG and quite a challenge to analyze.

The Vulkan approach of having the user explicitly provide the transitions ahead of time and having them baked into the command buffer, while not trying to insert them at runtime is most probably the biggest reason (jointly with offline shader compilation) why Vulkan applications are seeing the often quoted 60% overall-CPU-utilization reductions on mobile devices even when multithreading and worker-thread command buffer generation is accounted for.

Reply:
Yes, that's pretty much where the group is heading at the moment.

> I think a dependency graph of the resources used in the command buffers you will be processing for submission will be an enormous DAG and quite a challenge to analyze.

I'm not as pessimistic. There aren't that many resources that change their usage through the frame: say, a hundred render targets plus a bunch of buffers we write as UAV. This isn't an enormous DAG to analyze.

>The Vulkan approach of having the user explicitly provide the transitions ahead of time and having them baked into the command buffer, while not trying to insert them at runtime 

The good thing for WebGPU (contrary to OpenGL) is that we have command buffers. So the barriers *inside* command buffers will also need to be computed only once (either at recording, or the first submission, depending on implementation). It's just the barriers *between* command buffers that we'll need to compute and insert on every submission at runtime.

Reply:
The whole point of the "render pass" system added in the latest D3D12 is to help track and validate these transitions, Metal already does this tracking implicitly, and it's now a common tactic for engines to track them it themselves as well ( see https://www.gdcvault.com/play/1024656/Advanced-Graphics-Tech-Moving-to and http://32ipi028l5q82yhj72224m8j.wpengine.netdna-cdn.com/wp-content/uploads/2017/03/GDC2017-D3D12-And-Vulkan-Done-Right.pdf ).

Reply:
well yes, so since engines do it themselves.... maybe you should let them and make your lives easier?

Reply:
There is great deal of historical context to this discussion. Basically, since the browser implementations have to track the lifetime and usage of resources *for validation* anyway, it's not a big step to derive the barriers from this info.

Reply:
lifetime resource tracking is easiest, instead of deleting (where the usual JS GC would do) put it on a list with an associated API fence inserted after last API use (or instead of where you'd want the delete)... I do that already.

barriers are easy (in OpenGL you basically slam a glMemoryBarrier before the first read from a modified resource)

begin/end dependencies are a bit harder, and I would really like to see an implementation that is able to insert them and validate them before command submit time so we can benefit from pre-validated and pre-compiled command buffers like we go in D3D12 and VK.